presto-main/src/main/java/com/facebook/presto/operator/SingleChannelGroupByHash.java

```diff
@@ -0,0 +1,228 @@
+
+    private Type type;
+
+    // the hash table from values to groupIds
+    private BlockBuilder values;
+    private IntBigArray groupIds;
+    private IntBigArray hashPositions;
+
+    // groupId for the null value
+    private int nullGroupId = -1;
+
+    // reverse index from the groupId back to the value
+    private final BlockBuilder valuesByGroupId;
+
+    // the occurrence of each groupId
+    private final LongBigArray countsByGroupId;
+
+    private int nextGroupId;
+
+    public SingleChannelGroupByHash(Type type, int expectedSize)
+    {
+        this.type = type;
+
+        checkArgument(expectedSize > 0, "expectedSize must be greater than zero");
+
+        int hashSize = arraySize(expectedSize, FILL_RATIO);
+
+        maxFill = calculateMaxFill(hashSize);
+        mask = hashSize - 1;
+        values = this.type.createBlockBuilder(new BlockBuilderStatus(), hashSize, expectedValueSize(type, hashSize));
+        groupIds = new IntBigArray(-1);
+        groupIds.ensureCapacity(hashSize);
+        hashPositions = new IntBigArray(-1);
+        hashPositions.ensureCapacity(hashSize);
+
+        valuesByGroupId = this.type.createBlockBuilder(new BlockBuilderStatus(), hashSize, expectedValueSize(type, hashSize));
+        countsByGroupId = new LongBigArray();
+        countsByGroupId.ensureCapacity(hashSize);
+    }
+
+    public SingleChannelGroupByHash(Slice serialized, Type type, int expectedSize)
+    {
+        this(type, expectedSize);
+        checkNotNull(serialized, "serialized is null");
+        deserialize(serialized);
+    }
+
+    public long getEstimatedSize()
+    {
+        return groupIds.sizeOf() +
+                values.getRetainedSizeInBytes() +
+                valuesByGroupId.getRetainedSizeInBytes() +
+                countsByGroupId.sizeOf() +
+                hashPositions.sizeOf();
+    }
+
+    public int getGroupCount()
+    {
+        return nextGroupId;
+    }
+
+    public Block getValuesByGroupId()
+    {
+        return valuesByGroupId.build();
+    }
+
+    public LongBigArray getCountsByGroupId()
+    {
+        return countsByGroupId;
+    }
+
+    private void deserialize(Slice serialized)
+    {
+        Block block = TypeUtils.readStructuralBlock(serialized);
+        for (int i = 0; i < block.getPositionCount(); i += 2) {
+            putIfAbsent(i, block, block.getLong(i + 1, 0));
+        }
+    }
+
+    public Slice serialize()
+    {
+        Block values = valuesByGroupId.build();
+        BlockBuilder blockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus(), values.getSizeInBytes() + (int) countsByGroupId.sizeOf());
+        for (int i = 0; i < nextGroupId; i++) {
+            type.appendTo(values, i, blockBuilder);
+            BIGINT.writeLong(blockBuilder, countsByGroupId.get(i));
+        }
+        return buildStructuralSlice(blockBuilder);
+    }
+
+    public int putIfAbsent(int position, Block block, long count)
+    {
+        if (block.isNull(position)) {
```

Comment:
Since this will only be used for histograms, and you could never have NULL in a map key (and therefore never in the result of a histogram), you can remove all the NULL handling